### DISCUSSION TITLE
Краткий гайд для установки на Padavan / Entware

### DISCUSSION THREAD

**Discussion by votarum** [Anonym-tsk/nfqws-keenetic#76] [Делимся опытом]
1. Прошивка должа быть собрана с необходимыми модулями. Для этого в конфиг сборки добавляем:
```
CONFIG_NETFILTER_XT_TARGET_NFQUEUE=m
CONFIG_NETFILTER_NETLINK_QUEUE=m
CONFIG_IP_NF_QUEUE=m
```

2. После загрузки роутера подключаем модули xt_NFQUEUE и nfnetlink_queue:
```
modprobe xt_NFQUEUE
modprobe nfnetlink_queue
```
проверить, что они загрузились можно командой `lsmod`

3. Устанавливаем Entware, во внутренню память или на USB-флэшку.

4. Устанавливаем nfqws по [инструкции](https://github.com/Anonym-tsk/nfqws-keenetic) для Keenetic и Entware.

Чтобы после перезагрузки nfqws нормально стартовал, в пользовательский скрипт "Выполнить перед инициализацией маршрутизатора" добавляем:
```
modprobe xt_NFQUEUE
modprobe nfnetlink_queue
```
Чтобы после переподключения WAN/VPN nfqws работал, в пользовательский скрипт "Выполнить после перезапуска правил брандмауэра" добавляем:
```
killall nfqws
/opt/etc/init.d/S51nfqws start
```

**Comment by 4Net-Paladin** [comment]
Добавлю контекста.
Собирать актуальную прошивку Padavan можно онлайн, на серверах Github. [Инструкция по сборке находится здесь](https://github.com/shvchk/padavan-builder-workflow/blob/main/README.ru.md)

Эти самые три строчки:

**CONFIG_NETFILTER_XT_TARGET_NFQUEUE=m
CONFIG_NETFILTER_NETLINK_QUEUE=m
CONFIG_IP_NF_QUEUE=m**

добавляются в **build.config** перед запуском сборки прошивки

А вот [здесь дана инструкция по правильной заливке готовой прошивки в роутер](https://4pda.to/forum/index.php?act=findpost&pid=115912586&anchor=Spoil-115912586-5). Лучше не пренебрегать этими рекомендациями и не забывать чистить кэш браузера и куки после перепрошивки.

**Reply by tipaBoss** [reply]
Какие могут быть минусы, если сразу собрать прошивку с nfqws?
Флагом CONFIG_FIRMWARE_INCLUDE_NFQWS=y
В этом случае прямо из веб-морды можно настраивать.

**Reply by kotovasia5120** [reply]
где про флаги padavan почитать?

**Reply by kotovasia5120** [reply]
> "флаги" - это строчки из конфига, "почитайте" в репозитории

искал не нашел

**Reply by votarum** [reply]
не нашли что? репозиторий? конфиг в нем?

**Reply by kotovasia5120** [reply]
> е нашли что? репозиторий? конфиг в нем?

разъяснение значений флагов

**Reply by votarum** [reply]
> Какие могут быть минусы, если сразу собрать прошивку с nfqws?
> Флагом CONFIG_FIRMWARE_INCLUDE_NFQWS=y

Придется пересобирать прошивку и перепрошивать роутер, когда потребуется обновить nfqws. Но есть и плюс - не требуется Entware, будет работать на устройствах с 16М flash без USB.

p.s. даже в роутер с 8M flash без USB удалось запихнуть

**Comment by votarum** [comment]
Установка Entware во внутренню память (для устройств с NAND 128M):

```
cat /proc/mtd #убедится, что rwfs это mtd11
ubiformat /dev/mtd11
ubiattach -p /dev/mtd11 -d 11
ubimkvol /dev/ubi11 -m -N user
nvram set mtd_rwfs_mount=1
nvram set optw_enable=2
nvram commit
reboot
```
```
mkdir /media/mtd_rwfs/opt
reboot
```
```
opkg.sh
```

**Comment by artem0711** [comment]
Собираю последний Padavan-ng из [репозитория](https://gitlab.com/hadzhioglu/padavan-ng) для N56U А1 через padavan-builder-workflow.
Модули в build.config добавил:
```
CONFIG_NETFILTER_XT_TARGET_NFQUEUE=m
CONFIG_NETFILTER_NETLINK_QUEUE=m
CONFIG_IP_NF_QUEUE=m
```
Через modprobe два модуля подгружаются, а при старте nfqws просит ещё xt_multiport:
```
/opt/home/admin # /opt/etc/init.d/S51nfqws start
Cannot find xt_multiport.ko module
we have 3 user defined desync profile(s) and default low priority profile 0
Loading hostlist /opt/etc/nfqws/user.list
loading plain text list
Loaded 34 hosts from /opt/etc/nfqws/user.list
Loading hostlist /opt/etc/nfqws/user.list
loading plain text list
Loaded 34 hosts from /opt/etc/nfqws/user.list
Started NFQWS service
```
Обход блокировок при это работает. На сколько критично отсутствие подгруженного модуля xt_multiport.ko? Как-то его можно добавить в прошивку Padavan?

**Reply by votarum** [reply]
multiport уже есть в iptables, ничего не нужно добавлять

**Comment by StatCombo** [comment]
Добрый день, помогите пожалуйста нубу разобраться с установкой ентвари на внешний диск на прошивке падаван, собрал с нужными параметрами, прошил а дальше не понятно, как ставить nfqws его настраивать итд, и к сожалению у нас тут блочат все ресурсы, где это обсуждается... надоело сидеть без Ю... заранее спасибо за помощь

**Reply by DJ-URAN** [reply]
А что конкретно?

**Reply by StatCombo** [reply]
> А что конкретно?

Да пошагово как настраивается nfqws
Как установить её на узб флеш, чтобы нанд не портить, и как логи отключать...

**Reply by votarum** [reply]
с мифами идите к тем, кто вам их "подарил", nfqws ничего не портит

**Reply by StatCombo** [reply]
> с мифами идите к тем, кто вам их "подарил", nfqws ничего не портит

С мифами понятно, а по делу есть что сказать?

**Reply by votarum** [reply]
"По делу" все уже сказано. Установку Entware на USB в падаване только ленивый не нагуглит. Дальнейшая инструкция - на [самом видном месте](https://github.com/Anonym-tsk/nfqws-keenetic).

**Reply by StatCombo** [reply]
> "По делу" все уже сказано.<
Так бинарь от кинетика можно юзать на ми3ж?

**Reply by votarum** [reply]
на падаване, который собран, как написано выше

**Reply by StatCombo** [reply]
> на падаване, который собран, как написано выше

Так точно, собран именно так как выше сказано, я внизу конфига добавил 3 строчки и собрал

**Reply by StatCombo** [reply]
> "По делу" все уже сказано. Установку Entware на USB в падаване только ленивый не нагуглит. Дальнейшая инструкция - на [самом видном месте](https://github.com/Anonym-tsk/nfqws-keenetic).

Действительно, всё оказалось куда проще, чем я думал.... В любом случае спасибо за наводку на цепОчку!

**Reply by Snesssh** [reply]
Так блин) я замучался, можете поделиться ссылкой на установку entware? Ставлю на asus ac51u. Или ссылку как это примерно работает. Получается надо еще read only mode снять?

**Reply by Goshathome** [reply]
Создай пустую папку opt в корне на флешке. Воткни флешку в роутер и перезагрузи его. Прошивка сама всё поставит. По крайней мере так на прошивках Padavan
